### PR TITLE
SFM: Always consider radial distortion

### PR DESCRIPTION
**Pull request by flanggut**
- Add an undistortion function for image features and use corrected features for track triangulation.
- If possible restore initial outlier tracks for viewports after intrinsics have been optimized.

### CODE REVIEW

File: libs/sfm/bundler_incremental.cc [hunk 1]
```diff
@@ -188,6 +190,41 @@ Incremental::reconstruct_next_view (int view_id)
     return true;
 }
 
+void
+Incremental::try_restore_tracks_for_views (void)
+{
+  for (std::size_t i = 0; i < this->viewports->size(); ++i)
```
Comment (nmoehrle) [inline libs/sfm/bundler_incremental.cc]: Did you use two spaces for indentation on purpose?

File: libs/sfm/bundler_common.cc [hunk 2]
```diff
@@ -266,5 +266,50 @@ load_survey_from_file (std::string const& filename,
...
+    {
+        // Guess initial interval upper and lower bound
+        double lbound = r2, ubound = r2;
+        while (distort_radius(lbound, params) > r2)
+            lbound /= 1.05;
+        while (distort_radius(ubound, params) < r2 )
+            ubound *= 1.05;
+
+        // Perform a bisection until epsilon accuracy is reached
+        double const epsilon = 1e-10;
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]: You could use std::numeric_limits<double>::epsilon(), which is close to what you have, 1e-9.

File: libs/sfm/bundler_incremental.cc [hunk 3]
```diff
@@ -188,6 +190,41 @@ Incremental::reconstruct_next_view (int view_id)
...
+            || viewport.track_ids[feature_id] >= 0)
+            continue;
+        math::Vec3f const& pos3d = this->tracks->at(track_id).pos;
+        math::Vec2f const& pos2d = undistort_feature(
+            features.positions[feature_id], viewport.radial_distortion,
+            viewport.focal_length);
+
+        math::Vec4d const pos3dh(pos3d[0], pos3d[1], pos3d[2], 1.0);
+        math::Matrix<double, 3, 4> P;
+        viewport.pose.fill_p_matrix(&P);
```
Comment (nmoehrle) [inline libs/sfm/bundler_incremental.cc]: I would help the compiler here and move the P matrix out of the loop :-)

File: libs/sfm/bundler_common.cc [hunk 4]
```diff
@@ -266,5 +266,50 @@ load_survey_from_file (std::string const& filename,
     in.close();
 }
 
+/* ---------------------- Feature undistortion -------------------- */
+
+namespace {
+    double distort_radius (double const r2, float const* params)
+    {
+        double coeff = 1.0 + r2 * params[0] + r2 * r2 * params[1];
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]: Can you add a comment what you're computing in this function? I don't understand it. According to this formula, we return `r^2 * (1 + k0 r^2 + k2 r^4)^2` which contains the radius to the power of 8?!?

File: libs/sfm/bundler_incremental.cc (hunk 1 above)
Comment (flanggut) [reply libs/sfm/bundler_incremental.cc]: whoops :D

File: libs/sfm/bundler_common.cc [hunk 5]
```diff
@@ -266,5 +266,50 @@ load_survey_from_file (std::string const& filename,
...
+        return r2 * coeff * coeff;
+    }
+
+    double solve_undistorted_radius (double const r2, float const* params)
+    {
+        // Guess initial interval upper and lower bound
+        double lbound = r2, ubound = r2;
+        while (distort_radius(lbound, params) > r2)
+            lbound /= 1.05;
+        while (distort_radius(ubound, params) < r2 )
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]: remove space before bracket here, and two more times below.

File: libs/sfm/bundler_common.cc [hunk 6]
```diff
@@ -266,5 +266,50 @@ load_survey_from_file (std::string const& filename,
...
+namespace {
+    double distort_radius (double const r2, float const* params)
+    {
+        double coeff = 1.0 + r2 * params[0] + r2 * r2 * params[1];
+        return r2 * coeff * coeff;
+    }
+
+    double solve_undistorted_radius (double const r2, float const* params)
+    {
+        // Guess initial interval upper and lower bound
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]: Is there a formula to compute this close form, instead of bisecting it?

File: libs/sfm/bundler_common.cc [hunk 7]
```diff
@@ -266,5 +266,50 @@ load_survey_from_file (std::string const& filename,
...
+}
+
+math::Vec2f
+undistort_feature (math::Vec2f const& f, float const* params,
+    float const focal_length)
+{
+    // Convert to camera coords
+    double const r2 = f.square_norm() / MATH_POW2(focal_length);
+    double const undist_radius = solve_undistorted_radius(r2, params);
+    double const scale = (r2 == 0) ? 1. : std::sqrt(undist_radius / r2);
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]:
```
double scale = 1.0;
if (r2 != 0.0)
{
    double const undist_radius = ...
    scale = ...
}
```

This way we avoid computing undist radius if `r2 == 0`.

File: libs/sfm/bundler_common.cc [hunk 8]
```diff
@@ -266,5 +266,50 @@ load_survey_from_file (std::string const& filename,
...
+    math::Vec2f res = f * scale;
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]: `return f * scale;`

File: libs/sfm/bundler_incremental.cc [hunk 9]
```diff
@@ -188,6 +190,41 @@ Incremental::reconstruct_next_view (int view_id)
     return true;
 }
 
+void
+Incremental::try_restore_tracks_for_views (void)
+{
+  for (std::size_t i = 0; i < this->viewports->size(); ++i)
+  {
+    Viewport & viewport = this->viewports->at(i);
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_incremental.cc]: No space before `&`

File: libs/sfm/bundler_common.cc [hunk 10]
```diff
@@ -266,5 +266,50 @@ load_survey_from_file (std::string const& filename,
     in.close();
 }
 
+/* ---------------------- Feature undistortion -------------------- */
+
+namespace {
+    double distort_radius (double const r2, float const* params)
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]: Should this function explicitly take the undistortion parameters, i.e., `k0` and `k2`, or whatever it is? This way it's easier to understand and you may implicitly convert those to doubles (i.e., take `double k0` and `double k2` as parameters).

Comment (flanggut) [reply libs/sfm/bundler_common.cc]: changed

File: libs/sfm/bundler_common.cc (hunk 4 above)
Comment (flanggut) [reply libs/sfm/bundler_common.cc]: It simply computes the squared distorted radius. I added a comment to clarify this.

File: libs/sfm/bundler_common.cc (hunk 6 above)
Comment (flanggut) [reply libs/sfm/bundler_common.cc]: added comment

Comment (flanggut) [reply libs/sfm/bundler_common.cc]: We would have to find and test all roots of a large polynomial, which is not practical.

File: libs/sfm/bundler_common.cc [hunk 11]
```diff
@@ -266,5 +266,56 @@ load_survey_from_file (std::string const& filename,
...
+
+        // Guess initial interval upper and lower bound
+        double lbound = r2, ubound = r2;
+        while (distort_squared_radius(lbound, k1, k2) > r2)
+            lbound /= 1.05;
+        while (distort_squared_radius(ubound, k1, k2) < r2)
+            ubound *= 1.05;
+
+        // Perform a bisection until epsilon accuracy is reached
+        while (std::numeric_limits<double>::epsilon() < ubound - lbound)
```
Comment (simonfuhrmann) [inline libs/sfm/bundler_common.cc]: you need std::abs()?

Comment (flanggut) [reply libs/sfm/bundler_common.cc]: The upper bound is always bigger than the lower bound.

**Pull request merged**